cms/djangoapps/contentstore/helpers.py

```diff
@@ -378,6 +381,17 @@ def _import_xml_node_to_parent(
         for child_node in child_nodes:
             _import_xml_node_to_parent(child_node, new_xblock, store, user_id=user_id)
 
+    # .. event_implemented_name: XBLOCK_DUPLICATED
+    if copied_from_block:
+        XBLOCK_DUPLICATED.send_event(
+            time=datetime.now(timezone.utc),
+            xblock_info=DuplicatedXBlockData(
+                usage_key=new_xblock.location,
+                block_type=new_xblock.location.block_type,
+                source_usage_key=new_xblock.copied_from_block,
+            )
+        )
```

Comment:
I'm not sure this is the right way to do this..

There's a few issues:

1. We never receive an `XBLOCK_CREATED` event for the new xblock that is created here, because the new xblock isn't constructed by [modulestore.create_item()](https://github.com/openedx/edx-platform/blob/cee42fd543306627119f884da60e631a0f9de486/xmodule/modulestore/mixed.py#L743-L770).
2. We do receive an `XBLOCK_UPDATED` event when [modulestore.update_item()](https://github.com/openedx/edx-platform/blob/cee42fd543306627119f884da60e631a0f9de486/xmodule/modulestore/mixed.py#L834) is called.
  However, we get an `XBLOCK_UPDATED` event every time the block is updated, not just on this initial paste, so we can't use it to signal tag copying.

and

3. If the source block has been deleted since it was copied into the clipboard, then there won't be any tags to copy over anymore.
   https://github.com/openedx/modular-learning/issues/203 says "2. When an XBlock is copied in Studio, copy its tags without including them in the OLX", so we'll need to resolve that here too.

Other ideas:

1. Send an `XBLOCK_CREATED` event here instead of `XBLOCK_DUPLICATED`.

   This makes more logical sense, but the event handler will need to fetch the xblock to get its `copied_from_block` field, because `XBLOCK_CREATED`'s data doesn't include this information like `XBLOCK_DUPLICATED` does.
   
   We can avoid this by adding `copied_from_block` field to [`XBlockData`](https://github.com/openedx/openedx-events/blob/fb6c3e153b20173bdec55fec4db648c1efa2124e/openedx_events/content_authoring/data.py#L74) (which is sent with `XBLOCK_CREATED` and `XBLOCK_UPDATED`).

2. Create a new `XBLOCK_PASTED` event and send that instead, with data like `XBLOCK_DUPLICATED`.